libgadget/fof.c

```diff
@@ -232,35 +232,46 @@ struct FOFPrimaryPriv {
 #define FOF_PRIMARY_GET_PRIV(tw) ((struct FOFPrimaryPriv *) (tw->priv))
 
 static int
-HEADl(int stop, int i, TreeWalk * tw)
+HEADl(int stop, int i, int locked, int * Head)
 {
     int r;
     if (i == stop) {
         return -1;
     }
 //    printf("locking %d by %d in HEADl stop = %d\n", i, omp_get_thread_num(), stop);
-    lock_particle(i);
+    /* Try to lock the particle. Wait on the lock if it is less than the already locked particle,
```

Comment:
I think we may be able to add an optimization here (as done in your other PR), to advance towards the root without locking first. Because we only ever grow the tree in depth I think it should work; but perhaps leave this for the future?

Reply:
I thought about this, but I couldn't immediately figure out how to get the lock retry logic written in an clear and obviously correct way. I think that with a splay it shouldn't lead to a big speedup: the tree should be O(1) deep so we don't lock many particles. In the other PR it was more important because there was no splay.

Reply:
Indeed. Let's have comment to note the reason this apparently obvious optimization is skipped.